### PR TITLE
execution: fix different error code from MySQL when inserting incorrect bigint value

### LINKED ISSUE #17832: Different error code from MySQL when inserting incorrect bigint value
### 1. Minimal reproduce step (Required)

````
drop table if exists t0;create table t0 (col0 VARCHAR(0), col1 TINYINT, col2 TEXT, col3 BLOB, col4 BIGINT);insert into t0 values (NULL, '-128', 't', 'b', '2 ^ 63 - 1');
````

### 2. What did you expect to see? (Required)

From both errors it is clear I shouldn't use `2^63` in the value. But I'm confused with the meaning of the error.

````
mysql> drop table if exists t0;create table t0 (col0 VARCHAR(0), col1 TINYINT, col2 TEXT, col3 BLOB, col4 BIGINT);insert into t0 values (NULL, '-128', 't', 'b', '2 ^ 63 - 1');
Query OK, 0 rows affected (0.00 sec)

Query OK, 0 rows affected (0.02 sec)

ERROR 1265 (01000): Data truncated for column 'col4' at row 1
````
### 3. What did you see instead (Required)

````
-- MySQL 5.7
mysql> drop table if exists t0;create table t0 (col0 VARCHAR(0), col1 TINYINT, col2 TEXT, col3 BLOB, col4 BIGINT);insert into t0 values (NULL, '-128', 't', 'b', '2 ^ 63 - 1');
Query OK, 0 rows affected (0.02 sec)

Query OK, 0 rows affected (0.01 sec)

ERROR 1366 (HY000): Incorrect bigint value: '2 ^ 63 - 1' for column 'col4' at row 1

-- MySQL 8
mysql> drop table if exists t0;create table t0 (col0 VARCHAR(0), col1 TINYINT, col2 TEXT, col3 BLOB, col4 BIGINT);insert into t0 values (NULL, '-128', 't', 'b', '2 ^ 63 - 1');
Query OK, 0 rows affected (0.03 sec)

Query OK, 0 rows affected (0.08 sec)

ERROR 1265 (01000): Data truncated for column 'col4' at row 1
````

### 4. Affected version (Required)

````
commit 8369ffd500f3fb235d8b584ac4298b2e59d8db55 (HEAD -> master, origin/master, origin/HEAD)
Author: Soup <ilovesoup@gmail.com>
Date:   Tue May 26 15:38:29 2020 +0800
````

### PR DESCRIPTION
**Pull request by TszKitLo40**
What problem does this PR solve?

Issue Number: close #17832  

Problem Summary:
fix different error code from MySQL when inserting incorrect bigint value

What's Changed:
Return types.ErrTruncated when the data is truncated in the cast value function.  

### Check List 

Tests 

- Manual test (Same steps in the issue)

### Release note 

- fix different error code from MySQL when inserting incorrect bigint value

### CODE REVIEW

**Review by wjhuang2016 — COMMENTED** [review]
Please add a unit test for this change.

**Comment by TszKitLo40** [comment]
The error is throwed in the `StrToFloat` in `conver.go`, because '2 ^ 63 - 1' does not have valid float prefix. So should we make 2 ^ 63 - 1 pass the check of `getValidFloatPrefix`  and check if it is overflow and throw the `WarnDataTruncated` error? @wjhuang2016 @SunRunAway

**Comment by TszKitLo40** [comment]
My commit failed to pass the tests. I have no idea how to know whether the error should be thrown between `ErrWarnDataTruncated` and `ErrTruncatedWrongValueForField`

**Comment by SunRunAway** [comment]
The failed test case says,
```
[2020-09-09T09:27:58.004Z] ----------------------------------------------------------------------

[2020-09-09T09:27:58.004Z] FAIL: executor_test.go:2915: testSuite.TestEmptyEnum

[2020-09-09T09:27:58.004Z] 

[2020-09-09T09:27:58.004Z] executor_test.go:2922:

[2020-09-09T09:27:58.004Z]     c.Assert(terror.ErrorEqual(err, table.ErrTruncatedWrongValueForField), IsTrue, Commentf("err %v", err))

[2020-09-09T09:27:58.004Z] ... obtained bool = false

[2020-09-09T09:27:58.004Z] ... err [table:1265]Data truncated for column 'e' at row 1

[2020-09-09T09:27:58.004Z] 

[2020-09-09T09:27:58.004Z] 

[2020-09-09T09:27:58.004Z] ----------------------------------------------------------------------
```

Could you help fix it? @TszKitLo40

**Review by wshwsh12 — COMMENTED** [review]
Please resolve conflict.

**Comment by TszKitLo40** [comment]
> Please resolve conflict.

I have resolved it.

**Comment by TszKitLo40** [comment]
I found that the bug is that '2 ^ 31 - 1' would be truncated to 2. `ErrTruncatedWrongVal` error would be thrown and I change it to `ErrorTruncated`. I am not sure whether my idea is correct.

**Comment by TszKitLo40** [comment]
I found that the bug is that '2 ^ 31 - 1' would be truncated to 2. `ErrTruncatedWrongVal` error would be thrown and I change it to `ErrorTruncated`. I am not sure whether my idea is correct.

File: table/column.go [hunk 1]
```diff
@@ -177,7 +177,7 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		if err1 != nil {
 			logutil.BgLogger().Warn("Datum ToString failed", zap.Stringer("Datum", val), zap.Error(err1))
 		}
-		err = sc.HandleTruncate(types.ErrTruncatedWrongVal.GenWithStackByArgs(col.FieldType.CompactStr(), str))
+		err = sc.HandleTruncate(types.ErrTruncated.GenWithStackByArgs(col.FieldType.CompactStr(), str))
```
Comment (mmyj) [inline table/column.go:180]: Modifying the code here will cause upper-level misjudgment errors and cause other problems. We should find other ways to resolved this issue.

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Pull request closed without merging**